Frontend/HsParser.hs

```diff
@@ -265,17 +265,25 @@ pPrimPat =  PsdWildPat <$  symbol "_"
 
 -- Parse a parsed pattern (lowest priority)
 pPatMain :: PsM PsdPat
-pPatMain = chainl1 pPrimPat (pure PsdAppPat)
+pPatMain = chainl1 (chainl1 pPrimPat (pure PsdAppPat)) (pPatOr <$ symbol "||")
+
+-- Create a parsed or pattern
+pPatOr :: PsdPat -> PsdPat -> PsdPat
+pPatOr p1 p2 = PsdOrPat p1 p2
```

Comment:
You can remove `pPatOr` and use `PsdOrPat` directly. General piece of advice: only define "smart" constructors if (a) you really want to hide the data constructors, or (b) if they do some actual work (e.g. creating an arrow type given a list of types). If the implementation of the "smart" function is identical to the data constructor and you are not hiding the constructor then you are just making the code longer ;-)

Reply:
Oh yeah wow, I just realised how derpy this is 😄